common/src/main/java/io/netty/util/internal/UTF8Decoder.java

```diff
@@ -0,0 +1,174 @@
+     */
+    /* default */ static final int UTF8_ACCEPT = 0;
+
+    /**
+     * The state the decoder will be in if the decoder detects an invalid character.
+     */
+    /* default */ static final int UTF8_REJECT = 12;
+
+    /**
+     * The state table that maps bytes to character classes.
+     * <p>
+     * <p>
+     * Classes:
+     * <ul>
+     *     <li>00..7f -> 0</li>
+     *     <li>80..8f -> 1</li>
+     *     <li>90..9f -> 9</li>
+     *     <li>a0..bf -> 7</li>
+     *     <li>c0..c1 -> 8</li>
+     *     <li>c2..df -> 2</li>
+     *     <li>e0..e0 -> 10</li>
+     *     <li>e1..ec -> 3</li>
+     *     <li>ed..ed -> 4</li>
+     *     <li>ee..ef -> 3</li>
+     *     <li>f0..f0 -> 11</li>
+     *     <li>f1..f3 -> 6</li>
+     *     <li>f4..f4 -> 5</li>
+     *     <li>f5..ff -> 8</li>
+     * </ul>
+     * </p>
+     */
+    /* default */  static final byte[] CHARACTER_CLASSES = new byte[] {
+            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+            1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9,
+            7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
+            8, 8, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
+            10, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 4, 3, 3, 11, 6, 6, 6, 5, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8,
+    };
+
+    /**
+     * Maps the state and the character class to the next state
+     */
+    /* default */ static final byte[] TRANSLATION_TABLE = new byte[] {
+            0, 12, 24, 36, 60, 96, 84, 12, 12, 12, 48, 72, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12,
+            12, 0, 12, 12, 12, 12, 12, 0, 12, 0, 12, 12, 12, 24, 12, 12, 12, 12, 12, 24, 12, 24, 12, 12,
+            12, 12, 12, 12, 12, 12, 12, 24, 12, 12, 12, 12, 12, 24, 12, 12, 12, 12, 12, 12, 12, 24, 12, 12,
+            12, 12, 12, 12, 12, 12, 12, 36, 12, 36, 12, 12, 12, 36, 12, 12, 12, 12, 12, 36, 12, 36, 12, 12,
+            12, 36, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12,
+    };
+
+    public static String decode(byte[] bytes) {
+        return decode(bytes, 0, bytes.length);
+    }
+
+    public static String decode(byte[] bytes, int offset, int length) {
+        int size = bytes.length;
+        if (MathUtil.isOutOfBounds(offset, length, size)) {
+            throw new IndexOutOfBoundsException(illegalIndex(offset, length, size));
+        }
+        UTF8Processor processor = new UTF8Processor(new char[length]);
+
+        for (int i = 0; i < length; i++) {
```

Comment:
AFAIK, the JIT [really likes this kind of loop](https://wiki.openjdk.java.net/display/HotSpot/PerformanceTechniques) because it knows that `i` is a constant.
Its also a really basic optimization to [move stuff out of the loop](https://en.wikipedia.org/wiki/Loop-invariant_code_motion)